### PR TITLE
Search ux improvements

### PR DESCRIPTION
**Pull request by FCalabria**
### PR Summary: 

UX improvements on search:
- Added clear button to search input
- The search term is kept in sync between the main search and the search form in the header
- When focusing the main search in mobile, it automatically scrolls to the top of the page

### Why are these changes introduced?

The improvements were suggested by @al-mcquiston in the context of a new feature (that will be on a different PR). For reference, here is the [figma](https://www.figma.com/file/HFKhDU5jNbnpR7U0JnsAur/Query-Suggestions-in-Storefront-Search?node-id=93%3A2521&t=AHxRw6HVpelQWflc-0) on that feature that includes the improvements on this PR.

### What approach did you take?

I created a common `search-form` component to avoid duplicating code between the header and the main search. The input sync functionality is developed on the main search because the only use case where more than one search input is visible would be with that one present.

### Other considerations

You might want to review per commit, since I did each improvement on a different one.

### Visual impact on existing themes

New clear button on search inputs:

Hidden when the input is empty

<details>
<summary>Auto scroll when focusing main search</summary>

</details>

### Demo links

- [Store](https://mcicq3997fukw1yo-45840990230.shopifypreview.com)
- [Editor](https://os2-demo.myshopify.com/admin/themes/137451733014/editor)

### Checklist
- [x] Added PR summary for [release notes](https://themes.shopify.com/themes/dawn/styles/default#ReleaseNotes)
- [ ] Requested review from UX (Only for changes that are affecting the experience or perceivable visual details)
- [ ] Created a ticket for the [help.shopify.com](https://help.shopify.com) documentation team about updates to theme settings. (Internal-only task)
- [x] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [x] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [x] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [x] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [x] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

### CODE REVIEW

File: assets/component-search.css [hunk 1]
```diff
@@ -1,32 +1,71 @@
...
+  content: "";
+  position: absolute;
+  right: 0;
 }
 
-.search__button:focus {
+.search__button:focus-visible,
+.reset__button:focus-visible,
+.search__button:focus,
+.reset__button:focus {
```
Comment (ludoboludo) [inline assets/component-search.css:28]: One thing that comes to mind at quick glance is that we keep those separated `:focus-visible` declarations and `:focus` ones as a browser that doesn't support `focus-visible` like older versions of Safari won't read the whole thing if it doesn't support one selector. 
```suggestion
.search__button:focus-visible,
.reset__button:focus-visible {
  ...
}

.search__button:focus,
.reset__button:focus {
 ...
 }
```

**Review by sofiamatulis — COMMENTED** [review]
Nice improvements 🎉 Did a first pass

File: assets/component-search.css (hunk 1 above)
Comment (FCalabria) [reply assets/component-search.css:28]: TIL

File: locales/en.default.json [hunk 2]
```diff
@@ -36,7 +36,8 @@
       "previous": "Previous page"
     },
     "search": {
-      "search": "Search"
+      "search": "Search",
+      "reset": "Clear search term"
```
[the remark this answers is not in the document]
Comment (ludoboludo) [reply locales/en.default.json:40]: I think that sounds good 👍 I was trying to see what screen readers announce when you use a search input that has the `clear` button by default but it doesn't even announce it. [Video example](https://screenshot.click/18-24-57k3b-snsfz.mp4)

File: assets/search-form.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    // Prevent default so the form reset doesn't set the value gotten from the url on page load
+    event.preventDefault();
+    // Don't reset if the user has selected an element on the predictive search dropdown
+    if (!document.querySelector('a[aria-selected="true"')) {
+      this.input.value = '';
+    }
+  }
+}
+
+customElements.define('search-form', SearchForm);
```
[the remark this answers is not in the document]
Comment (FCalabria) [reply assets/search-form.js]: The `extend` is a very nice idea, I'll refactor that.

What do you think about the scope of each file, any concern there? I made them assuming the only moment that there is more than one search input is when the main search is visible, that's why most of the logic is on that file. But if we don't assume that, then most of the logic will move to `search-form.js`. That means loading a bigger file with the header (with code that is not really needed) but also we will have a more resilient code in case that assumption changes in the future. What do yo think?

[the remark this answers is not in the document]
Comment (ludoboludo) [reply assets/search-form.js]: That's a good point. Maybe it is better for now to keep it as you have it 🤔 Only load that script when actually needed. Since it's on the search page anyway it shouldn't be too impactful. 👍

File: assets/main-search.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    this.allSearchInputs = document.querySelectorAll('input[type="search"]');
+    this.setupEventListeners();
+  }
+
+  setupEventListeners () {
+    let allSearchForms = [];
+    this.allSearchInputs.forEach(input => allSearchForms.push(input.form))
+    if (allSearchForms.length > 1) {
+      allSearchForms.forEach(form =>
+        form.addEventListener('reset', this.onFormReset.bind(this))
```
Comment (ludoboludo) [inline assets/main-search.js:13]: Here, I think instead of adding another `reset` listener, we could add functionality to our original function `onSearchFormReset(event){}` from the `SearchForm` class using `super` (you can see it used in a few places already in our codebase)

I think this way should work: 
```javascript
onSearchFormReset(event) {
  super.onSearchFormReset(event);
  this.keepInSync('');
}
```

**Review by ludoboludo — COMMENTED** [review]
Hey 👋 a few more comments/ideas 👍

File: assets/main-search.js (hunk 4 above)
Comment (ludoboludo) [reply assets/main-search.js]: Ah right 👍 Makes sense.

File: assets/search-form.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+    const form = this.querySelector('form.search');
+    form.addEventListener('reset', this.onFormReset.bind(this));
+  }
+
+  onFormReset(event) {
+    // Prevent default so the form reset doesn't set the value gotten from the url on page load
+    event.preventDefault();
+    // Don't reset if the user has selected an element on the predictive search dropdown
+    if (!document.querySelector('a[aria-selected="true"')) {
```
Comment (ludoboludo) [inline assets/search-form.js:14]: I think this is meant to be `li[aria-selected="true"` 🤔 When I check it's the `li` element that is getting the `aria-selected` attribute updated. 

Though I'm not sure the check is working. Or maybe I'm not testing it properly. When I select one item in the dropdown using my keyboard then click on the reset button `X` it still empties the input value. 
[Video recording](https://screenshot.click/22-32-eld40-kvp2w.mp4)

File: assets/main-search.js [hunk 6]
```diff
@@ -5,37 +5,39 @@ class MainSearch extends SearchForm {
...
     let allSearchForms = [];
     this.allSearchInputs.forEach(input => allSearchForms.push(input.form))
-    if (allSearchForms.length > 1) {
-      allSearchForms.forEach(form =>
-        form.addEventListener('reset', this.onFormReset.bind(this))
-      );
+    this.input.addEventListener("focus", this.onInputFocus.bind(this));
+    if (allSearchForms.length < 2) {
+      return;
     }
```
Comment (ludoboludo) [inline assets/main-search.js:14]:
```suggestion
    if (allSearchForms.length < 2) return;
```

File: assets/search-form.js [hunk 7]
```diff
@@ -3,15 +3,20 @@ class SearchForm extends HTMLElement {
...
-    const form = this.querySelector('form.search');
-    form.addEventListener('reset', this.onFormReset.bind(this));
+    if (this.input) {
+      this.input.form.addEventListener('reset', this.onFormReset.bind(this));
+    }
+  }
+
+  shouldResetForm() {
+    return !document.querySelector('[aria-selected="true"] a')
   }
```
[the remark this answers is not in the document]
Comment (FCalabria) [reply assets/search-form.js:13]: I followed the pattern for the search button, that also does not submit when an element in the predictive search is active. Maybe @melissaperreault knows why this solution was adopted.
I don't know if it will be a lot of changes for one PR, but I could remove the `active` state and reset the list when the user tabs around, although that does not solve all the use cases because they could tab first and use the cursor later 🤔

File: sections/header.liquid [hunk 8]
```diff
@@ -318,8 +318,9 @@
             <div class="search-modal__content{% if settings.inputs_shadow_vertical_offset != 0 and settings.inputs_shadow_vertical_offset < 0 %} search-modal__content-top{% else %} search-modal__content-bottom{% endif %}" tabindex="-1">
               {%- if settings.predictive_search_enabled -%}
                 <predictive-search class="search-modal__form" data-loading-text="{{ 'accessibility.loading' | t }}">
-              {%- endif -%}
+              {%- else -%}
                 <search-form>
```
[the remark this answers is not in the document]
Comment (FCalabria) [reply sections/header.liquid]: Nice catch! Although it was easier to fix: I was missing a class and an unclosed html tag 🤦‍♀️

**Review by melissaperreault — CHANGES_REQUESTED** [review]
Hey team! Thanks for the hard work! I think we have the opportunity to improve the buyer experience still before shipping this change!

- [x] When clearing the query, we should remain in the "actively typing" more where the cursor is still in action and waiting for the next letter to be pressed. This would avoid buyer to continuously needing to reveal the keyboard every time they made a mistake or change their query term. [Video reference](https://screenshot.click/25-30-g0kfc-vc5g6.mp4)
- [x] The feedback as come in a little later but we are looking at differentiating the clear icon from the close icon. Stay tuned!

I'll continue to review and reply where you need more attention, thanks!

File: assets/component-search.css [hunk 9]
```diff
@@ -1,32 +1,70 @@
 .search__input.field__input {
```
Comment (sofiamatulis) [inline assets/component-search.css:1]: When tabbing the outline is cut off at the bottom at these 2 places:

https://screenshot.click/06-01-yjs3x-tc9kq.png

https://screenshot.click/06-01-3n5yb-rul2p.png

File: assets/component-search.css [hunk 10]
```diff
@@ -1,32 +1,70 @@
...
-  padding-right: 5rem;
+  padding-right: 9.8rem;
 }
 
 .search__button {
```
Comment (sofiamatulis) [inline assets/component-search.css:5]: After clearing the search term, the search results dont stop and keep loading:

https://screenshot.click/06-03-ykpwy-cj92h.mp4

Is this expected? Shouldnt it be paused?

File: config/settings_data.json [hunk 11]
```diff
@@ -133,9 +133,7 @@
               }
             }
           },
-          "block_order": [
```
Comment (sofiamatulis) [inline config/settings_data.json:136]: Were these changes intended? 👀 Can we revert them?

**Review by sofiamatulis — COMMENTED** [review]
This looks great :) Added a few small comments and will do a final pass in the code

File: assets/component-search.css (hunk 10 above)
Comment (FCalabria) [reply assets/component-search.css:5]: Nice suggestion!

File: config/settings_data.json [hunk 12]
```diff
@@ -168,11 +170,15 @@
               "type": "text",
               "settings": {
                 "heading": "Our mission",
-                "subtext": "<p>Share contact information, store details, and brand content with your customers.</p>"
+                "subtext": "<p>Share contact information, store details, and brand content with your customers.<\/p>"
```
Comment (ludoboludo) [inline config/settings_data.json:173]: we should revert this. You might have a formatter applied on save  🤔

File: sections/header.liquid [hunk 13]
```diff
@@ -312,51 +319,60 @@
...
+                          aria-autocomplete="list"
+                          autocorrect="off"
+                          autocomplete="off"
+                          autocapitalize="off"
+                          spellcheck="false"
+                        {%- endif -%}
+                      >
+                      <label class="field__label" for="Search-In-Modal-1">{{ 'general.search.search' | t }}</label>
+                      <input type="hidden" name="options[prefix]" value="last">
+                      <button type="reset" class="reset__button field__button {% if search.terms == blank %} hidden{% endif %}" aria-label="{{ 'general.search.reset' | t }}">
```
Comment (ludoboludo) [inline sections/header.liquid:348]:
```suggestion
                      <button type="reset" class="reset__button field__button{% if search.terms == blank %} hidden{% endif %}" aria-label="{{ 'general.search.reset' | t }}">
```

File: sections/header.liquid [hunk 14]
```diff
@@ -534,52 +550,61 @@
...
+                          aria-autocomplete="list"
+                          autocorrect="off"
+                          autocomplete="off"
+                          autocapitalize="off"
+                          spellcheck="false"
+                        {%- endif -%}
+                      >
+                      <label class="field__label" for="Search-In-Modal">{{ 'general.search.search' | t }}</label>
+                      <input type="hidden" name="options[prefix]" value="last">
+                      <button type="reset" class="reset__button field__button {% if search.terms == blank %} hidden{% endif %}" aria-label="{{ 'general.search.reset' | t }}">
```
Comment (ludoboludo) [inline sections/header.liquid:579]:
```suggestion
                      <button type="reset" class="reset__button field__button{% if search.terms == blank %} hidden{% endif %}" aria-label="{{ 'general.search.reset' | t }}">
```

**Review by ludoboludo — COMMENTED** [review]
Looking good 👍 ill test it on different browser and devices later today 👍

**Review by melissaperreault — DISMISSED** [review]
Thank you Paqui for the iteration! 🎉 

I am testing out and seems to work well for me!
I am just experience some difference for that following behaviour:

_When focusing the main search in mobile, it automatically scrolls to the top of the page_

- [ ] It's not always scrolling up for me on my Android Galaxy Samsung A70. (_Chrome version 108.0.5359.79, I updated my version and it's still happening_ ) There is an attempt at scrolling I can perceive but it seems like there is something preventing from scrolling further to the top. [Visual reference](https://screenshot.click/07-44-dgz9t-t77lg.png) of where it stops as I interact with the component 

Although, I am still able to search so for me it's more to look out for refinement for various device type. Otherwise UX wise it works as expected for the rest, great work team! 🎉

File: assets/main-search.js [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    }
+  }
+
+  keepInSync(value, target) {
+    this.allSearchInputs.forEach(input => {
+      if (input !== target) {
+        input.value = value;
+      }
+    });
+  }
```
Comment (ludoboludo) [inline assets/main-search.js:46]: One small thing with the sync but I don't think of it as a blocker: [video](https://screenshot.click/12-01-i30oe-hbv29.mp4)

**Review by ludoboludo — DISMISSED** [review]
Just shared something I noticed but otherwise works as I'd expect.

File: assets/main-search.js (hunk 15 above)
Comment (FCalabria) [reply assets/main-search.js:46]: I did some changes to fix that, take a look at my last commit and tell me what do you think!

**Review by sofiamatulis — APPROVED** [review]
Looks good 🎉 Thank you for these improvements 🚢
